File: xabber/src/main/java/com/xabber/android/data/extension/references/mutable/filesharing/FileInfo.java
package com.xabber.android.data.extension.references.mutable.filesharing;

import org.jivesoftware.smack.util.XmlStringBuilder;

public class FileInfo {

    public static final String FILE_ELEMENT = "file";

    public static final String ELEMENT_MEDIA_TYPE = "media-type";
    public static final String ELEMENT_NAME = "name";
    public static final String ELEMENT_DESC = "desc";
    public static final String ELEMENT_HEIGHT = "height";
    public static final String ELEMENT_WIDTH = "width";
    public static final String ELEMENT_SIZE = "size";
    public static final String ELEMENT_DURATION = "duration";

    private String mediaType;
    private String name;
    private String desc;
    private int height;
    private int width;
    private long size;
    private long duration;

    public FileInfo() {}

    public String getMediaType() {
        return mediaType;
    }

    public String getName() {
        return name;
    }

    public String getDesc() {
        return desc;
    }

    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }

    public long getSize() {
        return size;
    }

    public long getDuration() {
        return duration;
    }

    public void setMediaType(String mediaType) {
        this.mediaType = mediaType;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setDesc(String desc) {
        this.desc = desc;
    }

    public void setHeight(int height) {
        this.height = height;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public void setSize(long size) {
        this.size = size;
    }

    public void setDuration(long duration) {
        this.duration = duration;
    }

    public CharSequence toXML() {
        XmlStringBuilder xml = new XmlStringBuilder();
        xml.openElement(FILE_ELEMENT);
        if (getMediaType() != null && !getMediaType().isEmpty()) {
            xml.openElement(ELEMENT_MEDIA_TYPE);
            xml.append(getMediaType());
            xml.closeElement(ELEMENT_MEDIA_TYPE);
        }
        if (getName() != null && !getName().isEmpty()) {
            xml.openElement(ELEMENT_NAME);
            xml.append(getName());
            xml.closeElement(ELEMENT_NAME);
        }
        if (getHeight() > 0) {
            xml.openElement(ELEMENT_HEIGHT);
            xml.append(String.valueOf(getHeight()));
            xml.closeElement(ELEMENT_HEIGHT);
        }
        if (getWidth() > 0) {
            xml.openElement(ELEMENT_WIDTH);
            xml.append(String.valueOf(getWidth()));
            xml.closeElement(ELEMENT_WIDTH);
        }
        if (getSize() > 0) {
            xml.openElement(ELEMENT_SIZE);
            xml.append(String.valueOf(getSize()));
            xml.closeElement(ELEMENT_SIZE);
        }
        if (getDesc() != null && !getDesc().isEmpty()) {
            xml.openElement(ELEMENT_DESC);
            xml.append(getDesc());
            xml.closeElement(ELEMENT_DESC);
        }
        if (getDuration() > 0) {
            xml.openElement(ELEMENT_DURATION);
            xml.append(String.valueOf(getDuration()));
            xml.closeElement(ELEMENT_DURATION);
        }
        xml.closeElement(FILE_ELEMENT);
        return xml;
    }
}

File: xabber/src/main/java/com/xabber/android/data/extension/references/mutable/filesharing/FileInfo.kt
package com.xabber.android.data.extension.references.mutable.filesharing

import org.jivesoftware.smack.util.XmlStringBuilder

/**
 * Represents a <file> element at file-sharing reference according to [XEP-FILE]
 * The <file/> child element is used to describe the properties of the file, which are represented in the <media/> element.
 *
 * @link https://xabber.com/protocol/otb
 *
 * @param mediaType The IANA-registered content type of the file data. Must be not empty.
 * @param name The name of file. Must be not empty.
 * @param size The size of the image data in bytes. Must be not equal to zero.
 */
class FileInfo(val mediaType: String, val name: String, val size: Long) {

    var desc: String = ""
    var height = 0
    var width = 0
    var duration: Long = 0

    fun toXML() = XmlStringBuilder().apply {
        openElement(FILE_ELEMENT)

        optElement(ELEMENT_MEDIA_TYPE, mediaType)
        optElement(ELEMENT_NAME, name)
        optElement(ELEMENT_SIZE, size.toString())

        if (height > 0) {
            optElement(ELEMENT_HEIGHT, height.toString())
        }
        if (width > 0) {
            optElement(ELEMENT_WIDTH, width)
        }
        if (desc.isNotEmpty()) {
            optElement(ELEMENT_DESC, desc)
        }
        if (duration > 0) {
            optElement(ELEMENT_DURATION, duration)
        }

        closeElement(FILE_ELEMENT)
    }

    companion object {
        const val FILE_ELEMENT = "file"
        const val ELEMENT_MEDIA_TYPE = "media-type"
        const val ELEMENT_NAME = "name"
        const val ELEMENT_DESC = "desc"
        const val ELEMENT_HEIGHT = "height"
        const val ELEMENT_WIDTH = "width"
        const val ELEMENT_SIZE = "size"
        const val ELEMENT_DURATION = "duration"
    }

}